src/AuthenticationService.php

```diff
@@ -129,7 +129,7 @@ public function identifiers(): IdentifierCollection
     /**
      * Access the authenticator collection
      *
-     * @return \Authentication\Authenticator\AuthenticatorCollection|\Authentication\Authenticator\AuthenticatorInterface[]
+     * @return \Authentication\Authenticator\AuthenticatorCollection
```

Comment:
This must be kept for iteratiin to properly discover in IDE

Reply:
Claiming it's an array of `AuthenticatorInterface` instances is a lie. It then causes static analysers to rightly complain about use of collection methods like `$this->authenticators()->load()`.

Until your IDE comes up with a way to declare the type of individual item returned by an iterator you can use inline annotation above your `foreach` loop when iterating the result of this method as done on line `177`.

Reply:
Interesting. We do that also all the time in core
That's actually the correct way IMO to declare the iterable part of any object, also other objects and other collections with the concrete one.
We should rather keep this one line than having the need to inline fix this everywhere in code.

Reply:
Note: Semantically, the array part here is gone once you have another object attached.
Same with `\ArrayObject|\Foo[]`. If this was an array it would just be the last part.
Maybe we dont do these kind of magic iterations anymore in the future to prevent these smells in the first place ;)

Reply:
> Maybe we dont do these kind of magic iterations anymore in the future to prevent these smells in the first place ;)

Nah just the static analysers / IDEs need to get smarter and come up with a common way to annotating individual item type returned by iterators.

Psalm already seem to provide a way of annotating classes with implement `IteratorAggrerate` https://psalm.dev/docs/annotating_code/templated_annotations/#template-constraints (which is what `ObjectRegistry` does).  I am going to look into it again and try to annotate `ObjectRegistry` properly.

Reply:
Dunno, somethings code can be more readable and straightforward itself. Doesn't need all the foo magic :)
But either way, how can we now keep this in a way that it properly documents itself on the methods? Instead of forcing the user implementation code to inline annotate?
The way I fixed it seemed to do that good enough also for phpstan.

Reply:
The end user almost never needs to use this method or the collection directly, so let's just keep phpstan and pslam happy.

Reply:
Hmm, it feels like a step backwords here also for understanding and especially for working with the code..
But I wont object to moving forward given the general improvements of the PR.